### PR TITLE
Cleanups to AccessPath representation and implementation

### PR DESCRIPTION
**Pull request by msridhar**
This cleans up how we represent the root of an access path.  Previously we used a `Root` class which contained an `Element`, with a boolean field indicating if the root was actually the receiver (`this`, which has no `Element`).  Now, we just make the root a `@Nullable Element`, with `null` representing the receiver.  Since we check our code with NullAway now this should be safe 🙂 And it's a slightly more efficient representation.

Alongside that change, clean up some of the messy logic inside `AccessPath` for access path building.  In doing this cleanup, I noticed that our streams handling didn't correctly handle cases involving a stream of `Map`s.  Not super realistic, but in any case it is fixed now and a test is added.

### CODE REVIEW

File: nullaway/src/test/resources/com/uber/nullaway/testdata/NullAwayStreamSupportNegativeCases.java [hunk 1]
```diff
@@ -177,6 +177,11 @@ public Stream<Integer> apply(NullableContainer<String> container) {
             });
   }
 
+  private Stream<String> filterThenMapStreamOfMapsWithGet(
+      Stream<java.util.Map<String, Integer>> stream) {
+    return stream.filter(m -> m.get("hello") != null).map(n -> n.get("hello").toString());
```
Comment (msridhar) [inline nullaway/src/test/resources/com/uber/nullaway/testdata/NullAwayStreamSupportNegativeCases.java:182]: On the master branch we get a false positive warning for this code

File: nullaway/src/main/java/com/uber/nullaway/dataflow/NullnessStore.java [hunk 2]
```diff
@@ -235,15 +235,14 @@ public NullnessStore uprootAccessPaths(
...
+      if (element == null) {
         continue;
       }
-      Element element = ap.getRoot().getElement();
       for (LocalVariableNode fromVar : localVarTranslations.keySet()) {
         if (element.equals(fromVar.getElement())) {
           LocalVariableNode toVar = localVarTranslations.get(fromVar);
-          AccessPath newAP =
-              new AccessPath(new AccessPath.Root(toVar.getElement()), ap.getElements());
+          AccessPath newAP = AccessPath.switchRoot(ap, toVar.getElement());
```
Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/dataflow/NullnessStore.java:245]: This is a bug fix; before we would not preserve any map-get argument in the new access path

File: nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPath.java [hunk 3]
```diff
@@ -514,32 +527,30 @@ public boolean equals(Object o) {
     if (this == o) {
       return true;
     }
-    if (!(o instanceof AccessPath)) {
+    if (o == null || getClass() != o.getClass()) {
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPath.java:530]: Any significant differences here between using `instanceof` vs comparing `.getClass()`? I assume no support for subtyping, but `AccessPath` is `final`, so it should be ok?

File: nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPath.java [hunk 4]
```diff
@@ -554,7 +565,14 @@ public MapKey getMapGetArg() {
 
   @Override
   public String toString() {
-    return "AccessPath{" + "root=" + root + ", elements=" + elements + '}';
+    return "AccessPath{"
+        + "root="
+        + root
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPath.java:570]: I assume string concatenation works with `null`, but a) did we test this?, b) would it make more sense to make this `root == null ? "this" : root`?

File: nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessAnalysis.java [hunk 5]
```diff
@@ -165,7 +165,7 @@ private Set<Element> getNonnullReceiverFields(NullnessStore nullnessResult) {
     Set<AccessPath> nonnullAccessPaths = nullnessResult.getAccessPathsWithValue(Nullness.NONNULL);
     Set<Element> result = new LinkedHashSet<>();
     for (AccessPath ap : nonnullAccessPaths) {
-      if (ap.getRoot().isReceiver()) {
+      if (ap.getRoot() == null) {
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessAnalysis.java:168]: A small downside of this refactoring is that this is slightly less self-documenting. Maybe worth adding a one-line comment here noting that this is the case where the root is the method's receiver?

File: nullaway/src/main/java/com/uber/nullaway/dataflow/NullnessStore.java [hunk 6]
```diff
@@ -235,15 +235,14 @@ public NullnessStore uprootAccessPaths(
       Map<LocalVariableNode, LocalVariableNode> localVarTranslations) {
     NullnessStore.Builder nullnessBuilder = NullnessStore.empty().toBuilder();
     for (AccessPath ap : contents.keySet()) {
-      if (ap.getRoot().isReceiver()) {
+      Element element = ap.getRoot();
+      if (element == null) {
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/dataflow/NullnessStore.java:239]: Same as last file.

**Review by lazaroclapp — APPROVED** [review]
LGTM, but see some comments about comments :) (Also, apologies for the delay, was on PTO)

File: nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPath.java (hunk 4 above)
Comment (msridhar) [reply nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPath.java:570]: Agreed, will make this change

File: nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessAnalysis.java (hunk 5 above)
Comment (msridhar) [reply nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessAnalysis.java:168]: I can add a method `AccessPath.rootIsReceiver()` maybe, so it's still self-documenting?

Comment (lazaroclapp) [reply nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessAnalysis.java:168]: Makes sense to me :) Wouldn't help in the `!= null` check without sophisticated ensures/requires/contract logic we can't handle yet, but that case is already somewhat self-documenting :)

**Pull request merged**